### DISCUSSION TITLE
How to get Google profile picture

### DISCUSSION THREAD

**Discussion by premnkl** [keycloak/keycloak#13854] [Q&A] [answered]
Dear Team,
I have followed below configuration in keycloak console **to get the profile picture from google**. But picture attribute is not present in the output of keycloak object

keycloak version: **18.0.2**
we are using keycloak-js with React

1. Identity Providers → google→ Mappers Create
2. Identity Providers → google→ Mappers → name = picture
3. Identity Providers → google→ Mappers → Sync Mode Override = inherit
4. Identity Providers → google→ Mappers → Mapper Type = Attribute Importer
5. Identity Providers → google→ Mappers → Social Profile JSON Field Path = picture
6. Identity Providers → google→ Mappers → User Attribute Name = picture

Below configuration made for Client

1. Client → google→ Mappers Create
2. Client → google→ Mappers → name = picture
3. Client → google→ Mappers → Mapper Type = User Attribute
4. Client → google→ Mappers → User Attribute= picture
5. Client → google→ Mappers →Token Claim Name= picture
6. Client → google→ Mappers →Claim JSON Type= string
7. Client → google→ Mappers →Add to ID token = ON
8. Client → google→ Mappers →Add to access token = ON
9. Client → google→ Mappers →Add to userinfo = ON

**let us know, if there is any other configuration changes missed as part of this setup.**

Regards,
Prem

**Comment by lexcao** [comment, marked as answer]
Hi, @premnkl 
The configuration for the Google identity provider is correct.
For the client, there is already a built-in mapper for the `picture` attribute, could you please try that?

After That,
you need to re-login to see the changes.

**Reply by premnkl** [reply]
Dear @lexcao ,

Thanks for your response. _**picture attribute is not part of the built-in mapper.**_ 

Hence, I added a custom mapper. 

Attached image for your reference.

![built-in mapper](https://user-images.githubusercontent.com/108449384/185409688-06f630d4-1799-4b0d-8d09-1cb5c443d946.png)
![client-mapper](https://user-images.githubusercontent.com/108449384/185410139-c9d96c24-0b1c-435c-a337-b41d48ead29c.png)

**Reply by lexcao** [reply]
Hi, if you have a mapper with the same name as the built-in mapper, the built-in mapper would be not shown.
<img width="720" alt="image" src="https://user-images.githubusercontent.com/14901139/185451932-1e5fa4ab-0135-406c-bece-f6c0f2290701.png">

You could remove the custom `picture` mapper, then add a built-in `picture` mapper.
But I think it is not needed to do that.
From the built-in `picture` mapper, it is the same as your current customized one.
<img width="720" alt="image" src="https://user-images.githubusercontent.com/14901139/185452306-0cb58612-6e58-4a73-b0f5-34bb48491ef7.png">

---
> But picture attribute is not present in the output of keycloak object

My question is what does `the output of keycloak object`
The `picture` attribute is added to the token, refresh token, and id token.

**Reply by premnkl** [reply]
Dear @lexcao ,

PFB the output of the keycloak object
This output is read at the '**onAuthSuccess**' event

![keycloak_output](https://user-images.githubusercontent.com/108449384/185551234-6df07991-d6c8-4bbb-8da5-525093e287a5.png)

**Reply by lexcao** [reply]
Hi @premnkl 
Could you please help me debug with the following steps?

### Attempt 1
Check your user to see if there is an attribute with the name `picture`.
If so, it might be related to mapper of client.
If not, try attempt 2

### Attempt 2
I found the mapper `Sync Mode` is `inherit`.
What is your Google IDP config of sync mode?
* If it is `import` which means only the first login would update the field.
  * If so, we need to change to `force` to let `picture` attribute update to date.

**Reply by premnkl** [reply]
Dear @lexcao ,

Thanks for your continuous inputs.

by following steps in Attempt 2, picture attribute is available as part of keycloak object output.

Will check with the team internally and will share next update, if there is any issue, else will close the issue.

Thanks,
Prem

**Reply by premnkl** [reply]
@lexcao thanks a lot for your inputs

**Comment by EspenRoth** [comment]
Huge thank you - for anyone looking to capture the google user ID the json path is `sub` - important to set the sync mode to `force`
